ocaml/xapi/xapi_observer.ml

```diff
@@ -110,27 +110,34 @@ module Observer : ObserverInterface = struct
     Tracing.Export.Destination.File.set_compress_tracing_files enabled
 end
 
-let supported_components = ["xapi"; "xenopsd"]
+module Observer_component = struct
+  type t = Xapi | Xenopsd [@@deriving ord]
+
+  let to_string = function Xapi -> "xapi" | Xenopsd -> "xenopsd"
+
+  let of_string = function
+    | "xapi" ->
+        Xapi
+    | "xenopsd" ->
+        Xenopsd
+    | c ->
+        failwith (Printf.sprintf "Not a valid component: %s" c)
+end
```

Comment:
I recommend adding `let all = [Xapi; Xenopsd]`, and use `Components.all` instead of `supported_components`.

It's both shorter and lets the programmer know that it should contain all the values of the variant, maybe define it right after the variant so it's even more obvious it needs to be updated.

Alternatively, check whether [ppx_enumerate](https://github.com/janestreet/ppx_enumerate) is available